Adding quantized::linear function for pytorch mobile in c10

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26335 Unify Quantization APIs for add, pool and relu
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK
* #26211 Add support to call unpack for pytorch mobile quantized FC and Conv
* #26152 Adding quantized::conv2d function for pytorch mobile in c10
* **#26135 Adding quantized::linear function for pytorch mobile in c10**

Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17434885](https://our.internmc.facebook.com/intern/diff/D17434885)

supriyar committed:
Adding quantized::linear function for pytorch mobile in c10

Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

ljk53 commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-prepack.cc:
```diff
@@ -22,7 +22,7 @@ PrePackConvWeights::PrePackConvWeights(
       const size_t packed_weights_size =
           (sizeof(uint8_t) * kernel_size + sizeof(int32_t)) * c_stride;
-      packed_weights_ = aligned_alloc(16, packed_weights_size);
+      packed_weights_ = malloc(packed_weights_size);
```
Is this needed for this to work or is this some sort of optimization?

supriyar replied:
This was an optimization. It was originally using `malloc` in QNNPACK. But the android build fails with `aligned_alloc` so I switched back to `malloc`

ljk53 commented on cmake/Dependencies.cmake:
```diff
@@ -372,6 +373,17 @@ if(USE_NNPACK)
   endif()
 endif()
 
+IF(NOT TARGET clog)
```
Doesn't pytorch_qnnpack CMake have this? why do we need move it?

supriyar replied:
We add all other dependencies here except for clog. Without adding it I see an error when `clog.h` is included.
```
./aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/log.h:13:18: fatal error: clog.h: No such file or directory
 #include <clog.h>
                  ^
```

supriyar commented on CMakeLists.txt:
```diff
@@ -161,7 +161,7 @@ option(USE_OPENCL "Use OpenCL" OFF)
 option(USE_OPENMP "Use OpenMP for parallel code" ON)
 option(USE_PROF "Use profiling" OFF)
-option(USE_QNNPACK "Use QNNPACK (quantized 8-bit operators)" ON)
+option(USE_QNNPACK "Use QNNPACK (quantized 8-bit operators)" OFF)
```
I am currently testing QNNPACK on pytorch server builds too (x86) using the python tests. If I turn off `USE_QNNPACK` only for mobile then I won't be able to test it on x86.
I think the main issue is that we didn't rename the kernels when we forked QNNPACK so at runtime it picks up kernels from `third_party` instead.

dzhulgakov replied:
Hm... we'd need to fix it (rename the kernels), otherwise buck build will be broken too

supriyar replied:
Yeah Ashkan has a PR out for this. https://github.com/pytorch/pytorch/pull/26238

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -221,19 +224,122 @@ class QLinearInt8 final : public torch::OperatorKernel {
+    return output;
+  }
+#endif
+#if defined(USE_FBGEMM) || defined(USE_PYTORCH_QNNPACK)
```
I think you can simplify without this ifdef

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp:
```diff
@@ -117,18 +123,87 @@ class QLinearPackWeightInt8 final : public c10::OperatorKernel {
+      bias = bias_in.value();
+    } else {
+      bias = at::empty(rows_w, at::kFloat);
+      bias = at::quantize_linear(bias, 1.0, 0, kQInt32);
```
empty? maybe you wanted zeros?

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -2,11 +2,62 @@
+}
+#endif
 
+inline uint8_t QuantizeUint8(float scale, int32_t zero_point, float value) {
```
btw, this seems to be similar to fbgemm implementation used in quantize_val. Can they be the same?

supriyar replied:
Looks similar, but that function calls `fbgemm::Quantize` function. I wanted to avoid calling fbgemm functions from pytorch mobile build because I don't think they will be built on mobile.

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -675,6 +675,7 @@ class TestDynamicQuantizedLinear(TestCase):
         use_channelwise=st.booleans())
     def test_qlinear(self, batch_size, input_channels, output_channels,
                      use_bias, use_relu, use_multi_dim_input, use_channelwise):
+        torch.backends.quantized.engine = torch.fbgemm
```
same comment that you'd need a context manager probably

dzhulgakov replied:
I think it'd be even better to lift it into the test's base class to make sure that we run tests agains both engines without duplicating the tests

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

dzhulgakov reviewed (approved):
It looks good, but I'd fix the build mode issue before landing

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422588](https://our.internmc.facebook.com/intern/diff/D17422588)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422588](https://our.internmc.facebook.com/intern/diff/D17422588)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422588](https://our.internmc.facebook.com/intern/diff/D17422588)

[ghstack-poisoned]

supriyar committed:
Update on "Adding quantized::linear function for pytorch mobile in c10"


Summary:
This change adds the support to call QNNPACK using the refactored API for Linear operators (Fully Connected)
It also has certain cmake changes to enable builing and using pytorch_qnnpack inside aten
I have disabled USE_QNNPACK in CMakeLists.txt. Enabling it results in picking kernels from third_party/QNNPACK during runtime since the function names are the same.

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qlinear_qnnpack
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422588](https://our.internmc.facebook.com/intern/diff/D17422588)

[ghstack-poisoned]
